Frontend Bug Fixes & Backend Optimization / Issue #73

ryanm055 opened the pull request:
NOTE: This PR does not account for test cases as test cases will be done in another PR for Issue #68. Getting performance results using Silk, etc will be done when PR is accepted.

All requirements for Issue #73 have been accomplished.
1. Fixed Frontend Issues
-- Fixed UI issue where action (edit/delete) buttons are unreachable when character field inputs are too long.
-- Implemented form validation (only for supplies for now) to ensure text input does not surpass the desired max length specified in the models.py file. 

2. Fixed Backend Issues 
-- Made retrieving recent activities time and memory efficient.

ryanm055 committed:
In Process of Making Personal Fixes.

ryanm055 committed:
Finished Logging part.

ryanm055 committed:
Finished Personal Fixes

ryanm055 committed:
Made fixes for a bug. Will come back to it soon.

ryanm055 committed:
Quick UI fix.

ryanm055 committed:
.

ryanm055 committed:
Fixed test cases to suit new changes for supply.

ryanm055 committed:
Finished logging for good. Still haven't done performance / benchmark tests yet. Will happen soon.

ryanm055 committed:
making activity objects readable if needed.

Hamzenium committed:
Merge branch 'Prod' into ryan_branch_personal_fixes

Hamzenium commented:
I fixed a merge conflict related to Silk. The profiler should only be enabled through ENABLE_SILK = True in settings, since leaving it on by default adds unnecessary middleware overhead and slows down responses

ryanm055 committed:
Saving all changes.

ryanm055 committed:
Merge branch 'ryan_branch_personal_fixes' of https://github.com/yublueprint/blackcreek_dbms into ryan_branch_personal_fixes

Hamzenium commented on config/settings.py:
```diff
@@ -49,6 +50,7 @@
     'django.contrib.auth.middleware.AuthenticationMiddleware',
     'django.contrib.messages.middleware.MessageMiddleware',
     'django.middleware.clickjacking.XFrameOptionsMiddleware',
+    'silk.middleware.SilkyMiddleware',
```
@ryanm055 For these two files under config, you can discard the recent changes — the correct implementation already exists in the Prod branch.
```
if ENABLE_SILK:
    INSTALLED_APPS += ["silk"]
    MIDDLEWARE.insert(0, "silk.middleware.SilkyMiddleware")
```

Let’s keep it as it was before , we don’t want to introduce the Silk overhead in production. The middleware should only be added conditionally when ENABLE_SILK is set to True, as handled in the existing settings.

Hamzenium commented on config/settings.py:
```diff
@@ -39,6 +39,7 @@
     'django.contrib.messages',
     'django.contrib.staticfiles',
     'app',
+    'silk',
```
@ryanm055  we need to remove Silk from here as well. It’s better to just copy the original files from the Prod branch and paste them here to maintain consistency.

Hamzenium commented on app/logging/logging.py:
```diff
@@ -1,92 +1,137 @@
-        if (
-            os.path.exists(self.filename)
-            and os.path.getsize(self.filename) >= self.max_bytes
+# Making it so that it can do both files, whichever gets passed in.
```
Since we’re already using OOP in this file, we can make the code even cleaner and more maintainable by moving the helper functions (rotate_if_needed, formatLine) inside the class, and placing all related functions under the main Logger class (or wherever it fits best — I’ll let you decide)

ryanm055 committed:
Finally fixed all issues. Will make slides as a report soon. There was actually never a memory issue, when you're not in Debug, memory held from a page is quickly cleared after enteringa new page, as in Debug mode, memory will fill quickly as explained here: https://stackoverflow.com/questions/5299228/django-project-eats-memory

I decided to just use readlines and go backwards because it ensures there is no possible chunking error and that it's actually faster than the chunking method. My analysis was wrong as both were actually O(n^2). Regardless, speed is not an issue as when not in debug mode but have silk enabled, it actually only takes less than 400ms to load the entire full recent activities list. Memory not in debug mode doesn't go over 110 MiB and doesn't stack up.

ryanm055 committed:
Turned Silk and Debug off by default, also I used pip install django-debug-toolbar and pip install Pympler but Pympler memory tab is not working for now. But point is, there was never really an issue but It is definitely a lot more optimized than the previous version, and everything is more organized.

ryanm055 committed:
Fixing and updating old notes.

ryanm055 committed:
.

ryanm055 commented:
I have not finished yet, I have to investigate why the results for the old version are better than new version. But if you want to see the information I have, feel free to look at the new slides in making.

New report:https://docs.google.com/presentation/d/1wYPmTYUdl9LZG6hb__xbH9fhwTKVuxvi_C-UGfzwwXY/edit?usp=sharing

Hamzenium commented:
@ryanm055  Great job on the analysis, the presentation looks solid!
Just one improvement suggestion: refactor the OOP design so that helpers like rotate_if_needed and formatLine become private methods inside the Logger class (in Python, private methods typically start with a double underscore __method_name), and the inner activityObject is defined as a proper class.

Once that’s updated, we should be all set to finalize and merge the PR. Also, don’t forget to click “Resolve conversation” so the merge option becomes available.

Hamzenium commented:
Also @ryanm055 , the pagination point you mentioned is a good catch, we can include that in our upcoming sprint. Once the app scales in production and data volume increases, pagination will help prevent excessive memory usage.

Hamzenium commented on app/logging/logging.py:
We can make this as part of the class itself, for better OOP design

Hamzenium reviewed (commented):
OOP Design for helper functions

ryanm055 committed:
Proper OOP implementation fixed memory and speed issues!

Thank you!!!!!!

ryanm055 committed:
Memory panel partly works but not completely, so commenting out print statements for now.

ryanm055 committed:
Not adding debug toolbar and Pympler for now.

ryanm055 commented:
I believe everything is all done now, please take a look at new slides:
* https://docs.google.com/presentation/d/1wYPmTYUdl9LZG6hb__xbH9fhwTKVuxvi_C-UGfzwwXY/edit?usp=sharing

Hamzenium commented on config/settings.py:
```diff
@@ -24,10 +24,29 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
+ENABLE_SILK = False
+if (DEBUG):
```
Since DEBUG is set to True, ENABLE_SILK is automatically enabled as well, which introduces additional overhead. Should we set DEBUG to False instead?

ryanm055 replied:
When Debug is true, it enables other debugging tools. I think it's good to have ENABLE_SILK in there because sometimes you wanna just use Silk in isolation and nothing else running as other things can affect performance as mentioned in the slides.

Hamzenium replied:
That makes sense, good call. We can set it to False for now so no debugging tools add any overhead, that should be fine 👍

Hamzenium reviewed (dismissed):
Should we set DEBUG to False instead?

Hamzenium reviewed (approved):
Looks Good, Ship it!

Hamzenium commented:
```
# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True
ENABLE_SILK = False
if (DEBUG):
    ENABLE_SILK = True
 ```
So even though ENABLE_SILK was initially set to False, once DEBUG is True, it automatically gets overridden and becomes True again.

That’s why I suggested setting DEBUG = False, in production, we don’t want SILK (or any profiling tools) running since they add extra overhead. But it’s fine for now; we can leave it as is for development.

If you start your server and navigate to:
```
http://127.0.0.1:8000/silk
```
you’ll notice that Silk is running and accessible from your local environment.

ryanm055 committed:
Merge branch 'Prod' into ryan_branch_personal_fixes
